Add final blog for GitLab Plugin Modernization project (#6628)
jenkins-infra/jenkins.io: content/blog/2023/08/02/2023-08-23-gitlab-plugin-modernization-report.adoc

---

== Overview

. <<About Project>>
. <<Coding-Phase 1>>
. <<Coding-Phase 2>>
. <<Future Improvements>>
. <<Acknowledgments>>
. <<Useful Links>>

---

== About Project

Plugins grows old and becomes difficult to maintain with time. GitLab Plugin was one such plugin that needed some love and care. The GitLab Plugin Modernization project thus aimed to modernize the GitLab Plugin by the use of old (link:https://resteasy.dev[RESTEasy]) library with modern (link:https://github.com/gitlab4j/gitlab4j-api[GitLab4J-API]) library via Jenkins (link:https://plugins.jenkins.io/gitlab-api/[GitLab API plugin]), normalizing the GitLab plugin’s use of the GitLab API with regard to the (link:https://plugins.jenkins.io/gitlab-oauth/[GitLab Authentication]), (link:https://plugins.jenkins.io/gitlab-branch-source/[GitLab Branch Source]), and (link:https://plugins.jenkins.io/gitlab-logo/[GitLab Logo]) plugins. This modernization reduces future maintenance burden and improves the overall quality of the plugin by making it more lightweight due to reduced dependencies and enhanced documentation.

The migration was performed in the following 2 coding phases:

---

== Phase 1

> In this phase, the following tasks were completed:

Reviewer: ```suggestion
In this phase, the following tasks were completed:

```

* Migration of GitLab Plugin from RESTEasy to GitLab4J-API via the GitLab-API plugin, without affecting the webhook functionality of the plugin. (link:[PR])
* Interactive testing of partially migrated plugin.

The achievenments of this phase were demonstrated in the (link:https://youtu.be/W4eSVCTmqb8[Midterm Presentation])

---

== Phase 2

In this phase, the following tasks were completed:

> * Adapting webhooks implimentation to GitLab4J-API.(link:https://github.com/jenkinsci/gitlab-plugin/pull/1553[PR])

Reviewer: ```suggestion
* Adapting webhooks implementation to GitLab4J-API.(link:https://github.com/jenkinsci/gitlab-plugin/pull/1553[PR])
```

---
:layout: post

> :title: "GitLab Plugin Modernization"

Reviewer: ```suggestion
:title: "GSoC GitLab Plugin Modernization Project"
```

:tags:
- gsoc
- gsoc2023
- gitlab
- plugins
:author: harsh-ps-2003
:sig: gsoc
description: >

>   The GitLab Plugin Modernization is a project under Google Summer of Code 2023.

Reviewer: ```suggestion
  The GitLab Plugin Modernization is a project under Google Summer of Code (GSoC) 2023.
```

opengraph:
  image: /images/gsoc/opengraph.png
links:
  gitter: https://app.gitter.im/#/room/#jenkinsci_gitlab-plugin:gitter.im
---

The goal of this blog is to showcase the work done on GitLab Plugin during the Google Summer of Code 2023.

> For a detailed description of the project, please see the link:/projects/gsoc/2023/projects/gitlab-plugin-modernization/[project page].

Reviewer 2: Fixed in 9ef8ae06fd482cfa3e3ea260ab671d8ece3b1b2f

---

== Overview

. <<About Project>>
. Coding <<Phase 1>>
. Coding <<Phase 2>>
. <<Future Improvements>>
. <<Acknowledgments>>
. <<Useful Links>>

---

== About Project

> Plugins grows old and becomes difficult to maintain with time. GitLab Plugin was one such plugin that needed some love and care. The GitLab Plugin Modernization project thus aimed to modernize the GitLab Plugin by the use of old (link:https://resteasy.dev[RESTEasy]) library with modern (link:https://github.com/gitlab4j/gitlab4j-api[GitLab4J-API]) library via Jenkins (link:https://plugins.jenkins.io/gitlab-api/[GitLab API plugin]), normalizing the GitLab plugin’s use of the GitLab API with regard to the (link:https://plugins.jenkins.io/gitlab-oauth/[GitLab Authentication]), (link:https://plugins.jenkins.io/gitlab-branch-source/[GitLab Branch Source]), and (link:https://plugins.jenkins.io/gitlab-logo/[GitLab Logo]) plugins. This modernization reduces future maintenance burden and improves the overall quality of the plugin by making it more lightweight due to reduced dependencies and enhanced documentation.

Reviewer 3: ```suggestion
Plugins grow old and become difficult to maintain with time.
The GitLab Plugin was one such plugin that needed some love and care.
Therefore, the GitLab Plugin Modernization project aimed to modernize the GitLab Plugin by using an old (link:https://resteasy.dev[RESTEasy]) library with a modern (link:https://github.com/gitlab4j/gitlab4j-api[GitLab4J-API]) library via Jenkins using the link:https://plugins.jenkins.io/gitlab-api/[GitLab API plugin].
This includes normalizing the GitLab plugin’s use of the GitLab API with regard to the link:https://plugins.jenkins.io/gitlab-oauth/[GitLab Authentication], link:https://plugins.jenkins.io/gitlab-branch-source/[GitLab Branch Source], and link:https://plugins.jenkins.io/gitlab-logo/[GitLab Logo] plugins.
This modernization reduces future maintenance burden and improves the overall quality of the plugin, by making it more lightweight thanks to reduced dependencies and enhanced documentation.
```

suggestion to update formatting and content, as there were missing words from the sentences that help with readability. Also updated to use full [sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) formatting.

The migration was performed in the following 2 coding phases:

---

== Phase 1

In this phase, the following tasks were completed:

> * Migration of GitLab Plugin from RESTEasy to GitLab4J-API via the GitLab-API plugin, without affecting the webhook functionality of the plugin. (link:[PR])

Reviewer 3: ```suggestion
* Migration of GitLab Plugin from RESTEasy to GitLab4J-API via the GitLab-API plugin, without affecting the webhook functionality of the plugin.
```
The link could be inserted in the text instead of at the end of the sentence, but the link needs to be pointing to the intended destination.

> * Interactive testing of partially migrated plugin.

Reviewer 3: ```suggestion
* Interactive testing of the partially migrated plugin.
```
adding in a word for readability

> The achievenments of this phase were demonstrated in the (link:https://youtu.be/W4eSVCTmqb8[Midterm Presentation])

Reviewer 3: ```suggestion
The achievements of this phase were demonstrated in the link:https://youtu.be/W4eSVCTmqb8[Midterm Presentation].
```
suggestion for spelling, removing () from link as they are not necessary for links

Reviewer 2: Fixed in e15b65f2da02ed8b5cfba8cd6b2a38399a02714b and a5f9d1a441cc80c0bcdd02813a4a375328ef6e66

---

== Phase 2

> In this phase, the following tasks were completed:
> * Adapting webhooks implementation to GitLab4J-API.(link:https://github.com/jenkinsci/gitlab-plugin/pull/1553[PR])
> * Migrating the Proxy Settings from RESTEasy to GitLab4J-API.(link:https://github.com/jenkinsci/gitlab-plugin/pull/1566[PR])
> * Partially adapting the Unit and Integration tests.
> * Improving the documentation of the plugin.(link:https://github.com/jenkinsci/gitlab-plugin/pull/1556[PR])
> * Interactive testing of the fully migrated plugin.
> 
> The results of this phase will be showcased in in the Final Presentation in the future.

Reviewer 3: ```suggestion
In this phase, the following tasks were completed:

* link:https://github.com/jenkinsci/gitlab-plugin/pull/1553[Adapting webhooks implementation to GitLab4J-API].
* link:https://github.com/jenkinsci/gitlab-plugin/pull/1566[Migrating the proxy settings from RESTEasy to GitLab4J-API].
* Partially adapting the Unit and Integration tests.
* link:https://github.com/jenkinsci/gitlab-plugin/pull/1556[Improving the documentation of the plugin].
* Interactive testing of the fully migrated plugin.

The results of this phase will be showcased in in the final presentation in the future.
```
suggestion for formatting/making links part of the text by adding them to the in line text as opposed to after, removing () from links as they are not necessary.

---

== Future Improvements

> - As GitLab has discontinued the GitLab API V3, the plugin should only support GitLab API V4 and above. Thus the plugin will be updated to support GitLab API V4 and above only.
> - The plugin lacks the test coverage, thus the plugin's Unit, Integrations as well as Docker-based Tests will have to be updated.
> - GitLab 16.0 has introduced some breaking changes which will have to be handled in the plugin.

Reviewer 3: ```suggestion
- As GitLab has discontinued the GitLab API V3, the plugin should only support GitLab API V4 and above.
Therefore, the plugin will be updated to support GitLab API V4 and above only.
- The plugin lacks the test coverage.
This means the plugin's unit, integrations, and Docker-based tests will have to be updated.
- GitLab 16.0 has introduced some breaking changes which will have to be handled in the plugin.
```
suggestion for formatting and adjusting content a bit. If this changes the messaging in the second point, please let me know!

---

== Acknowledgments

> * My gratitude knows no bounds that I got the opportunity to contribute to such impactful project under Google Summer of Code with such kind mentors. I would like to thank my mentors, link:/blog/authors/krisstern/[Kris Stern] and link:/blog/authors/MarkEWaite/[Mark Waite], for their constant support and guidance throughout the project. They put in a lot of effort in interactively testing the migrated plugin and providing me with valuable feedback. They never let me be stuck in any difficulty for too long and always helped me in finding the solution to the problem. I would also like to thank link:/blog/authors/basil/[Basil Crow] for the guidance he provided me with during the initial phase of the project, which helped the team shape the project.

Reviewer 3: ```suggestion
My gratitude knows no bounds that I got the opportunity to contribute to such impactful project under Google Summer of Code with such kind mentors.

I would like to thank my mentors, link:/blog/authors/krisstern/[Kris Stern] and link:/blog/authors/MarkEWaite/[Mark Waite], for their constant support and guidance throughout the project.
They put in a lot of effort interactively testing the migrated plugin and providing me with valuable feedback.
They never let me be stuck in any problem for too long and always helped me find the solution to the issue.
I would also like to thank link:/blog/authors/basil/[Basil Crow] for the guidance he provided during the initial phase of the project, which helped the team shape the project.
```
suggestion for updating syntax/removing bullet points

Reviewer 2: Resolved in cb4d306cb923c36d8101cf4dce9d126c6448613f

> * Also, thanks to the org admins, link:/blog/authors/jmmeessen[Jean-Marc Meessen], link:/blog/authors/alyssat[Alyssa Tong], and link:/blog/authors/krisstern/[Kris Stern], for always keeping me and other contributors on our toes and assisting us with any blockers and concerns by organizing weekly stand-up calls and personal 1-to-1 meets every phase.

Reviewer: ```suggestion
* Also, thanks to the org admins, link:/blog/authors/jmmeessen[Jean-Marc Meessen], link:/blog/authors/alyssat[Alyssa Tong], link:/blog/authors/krisstern/[Kris Stern], and link:/blog/authors/gounthar/[Bruno Verachten], for always keeping me and other contributors on our toes and assisting us with any blockers and concerns by organizing weekly stand-up calls and personal 1-to-1 meets every phase.
```

> * Also, thanks to the org admins, link:/blog/authors/jmmeessen[Jean-Marc Meessen], link:/blog/authors/alyssat[Alyssa Tong], link:/blog/authors/krisstern/[Kris Stern], and link:/blog/authors/gounthar/[Bruno Verachten], for always keeping me and other contributors on our toes and assisting us with any blockers and concerns by organizing weekly stand-up calls and personal 1-to-1 meets every phase.

Reviewer 3: ```suggestion
I also want to share thanks to the org admins, link:/blog/authors/jmmeessen[Jean-Marc Meessen], link:/blog/authors/alyssat[Alyssa Tong], link:/blog/authors/krisstern/[Kris Stern], and link:/blog/authors/gounthar/[Bruno Verachten], for always keeping me and other contributors on our toes, assisting us with any blockers and concerns by organizing weekly stand-up calls, and scheduling personal 1-to-1 meets during every phase.
```
suggestion to remove bullet point and adjust some of the text

Reviewer 2: Resolved in cb4d306cb923c36d8101cf4dce9d126c6448613f

---

== Useful Links

- link:https://plugins.jenkins.io/gitlab-plugin/[GitLab Plugin]
- link:https://github.com/jenkinsci/gitlab-plugin[Plugin's GitHub Repository]   

> - Use the link:https://app.gitter.im/#/room/#jenkinsci_gitlab-plugin:gitter.im[Gitter channel] or link:https://community.jenkins.io[community.jenkins.io] in case you have any question(s) or feedback.

Reviewer: ```suggestion
- Use the link:https://matrix.to/#/#jenkinsci_gitlab-plugin:gitter.im[Gitter channel] or link:https://community.jenkins.io[community.jenkins.io] in case you have any question(s) or feedback.
```

> - Use the link:https://matrix.to/#/#jenkinsci_gitlab-plugin:gitter.im[Gitter channel] or link:https://community.jenkins.io[community.jenkins.io] in case you have any question(s) or feedback.

Reviewer: ```suggestion
- Use the link:++https://matrix.to/#/#jenkinsci_gitlab-plugin:gitter.im++[Gitter channel] or link:https://community.jenkins.io[community.jenkins.io] in case you have any question(s) or feedback.
```
